### PR TITLE
chore(w3c): modify traceparent property and add tracestate property to context

### PR DESCRIPTION
**Pull request by ZStriker19**
## Description
In order for tracecontext propagation style injection, we need to be able to create traceparent and tracestate values to inject. This PR adds support for traceparent and tracestate in the context object. 

## Design

The traceparent and tracestate values need to be held from an incoming request to injection of an outgoing request. The context object already does the same for the origin tag.

For future work, we can also use this traceparent in DBM propagation.

## Checklist
- [x] Followed the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) when writing a release note.
- [x] Add additional sections for `feat` and `fix` pull requests.
- [x] [Library documentation](https://github.com/DataDog/dd-trace-py/tree/1.x/docs) and/or [Datadog's documentation site](https://github.com/DataDog/documentation/) is updated. Link to doc PR in description.

## Reviewer Checklist
- [x] Title is accurate.
- [x] Description motivates each change.
- [ ] No unnecessary changes were introduced in this PR.
- [ ] Avoid breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [ ] Tests provided or description of manual testing performed is included in the code or PR.
- [ ] Release note has been added and follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines), or else `changelog/no-changelog` label added.
- [ ] All relevant GitHub issues are correctly linked.
- [ ] Backports are identified and tagged with Mergifyio.

### CODE REVIEW

File: ddtrace/context.py [hunk 1]
```diff
@@ -125,9 +132,56 @@ def _traceparent(self):
         # type: () -> str
         if self.trace_id is None or self.span_id is None:
             return ""
+        tp = self._meta.get(_TRACEPARENT_KEY)
```
Comment (mabdinur) [inline ddtrace/context.py:135]: Do we get the traceparent from the meta dictionary to avoid encoding the trace id multiple times? If so what's the expected performance improvement? This change make this function harder to read so I want to know if it's worth it.

File: ddtrace/context.py [hunk 2]
```diff
@@ -125,9 +132,56 @@ def _traceparent(self):
...
+        # type: () -> str
+        dd = ""
+        if self.sampling_priority:
+            dd += "s:{};".format(self.sampling_priority)
+        if self.dd_origin:
+            dd += "o:{};".format(self.dd_origin)
+        sampling_decision = self._meta.get(SAMPLING_DECISION_TRACE_TAG_KEY)
+        if sampling_decision:
+            # replace characters ",", "=", and characters outside the ASCII range 0x20 to 0x7E
+            dd += "t.dm:{};".format(re.sub(r",|=|[^\x20-\x7E]+", "_", sampling_decision))
```
Comment (mabdinur) [inline ddtrace/context.py:156]: do we have a test to validate this regex?

File: ddtrace/context.py [hunk 3]
```diff
@@ -125,9 +132,56 @@ def _traceparent(self):
...
+            dd += "t.usr.id:{};".format(self._meta.get(USER_ID_KEY))
+
+        # grab all other _dd.p values out of meta since we need to propagate all of them
+        for k, v in self._meta.items():
+            if isinstance(k, str) and k.startswith("_dd.p") and k not in [SAMPLING_DECISION_TRACE_TAG_KEY, USER_ID_KEY]:
+                # for key replace ",", "=", and characters outside the ASCII range 0x20 to 0x7E
+                # for value replace ",", ";", ":" and characters outside the ASCII range 0x20 to 0x7E
+                next_tag = "{}:{};".format(
+                    re.sub("_dd.p.", "t.", re.sub(r",|=|[^\x20-\x7E]+", "_", k)), re.sub(r",|;|:|[^\x20-\x7E]+", "_", v)
+                )
```
Comment (mabdinur) [inline ddtrace/context.py:169]: We do use the valid ascii regex in a few places. We can probably extract this logic into its own function/constant.

File: ddtrace/context.py [hunk 4]
```diff
@@ -125,9 +132,56 @@ def _traceparent(self):
...
+            # replace characters ",", "=", and characters outside the ASCII range 0x20 to 0x7E
+            dd += "t.dm:{};".format(re.sub(r",|=|[^\x20-\x7E]+", "_", sampling_decision))
+        # since this can change, we need to grab the value off the current span
+        usr_id_key = self._meta.get(USER_ID_KEY)
+        if usr_id_key:
+            dd += "t.usr.id:{};".format(self._meta.get(USER_ID_KEY))
+
+        # grab all other _dd.p values out of meta since we need to propagate all of them
+        for k, v in self._meta.items():
+            if isinstance(k, str) and k.startswith("_dd.p") and k not in [SAMPLING_DECISION_TRACE_TAG_KEY, USER_ID_KEY]:
```
Comment (mabdinur) [inline ddtrace/context.py:164]: In python2 has two text types: [unicode and str](https://uwpce-pythoncert.github.io/SystemDevelopment/unicode.html#strings-vs-unicode).  So `isinstance(u"some_tag_key", str)` will evaluate to False in python2. 

We can use `isinstance(k, six.string_types)` instead:

```suggestion
            if isinstance(k, six.string_types) and k.startswith("_dd.p") and k not in [SAMPLING_DECISION_TRACE_TAG_KEY, USER_ID_KEY]:
```

File: ddtrace/context.py [hunk 5]
```diff
@@ -125,9 +132,56 @@ def _traceparent(self):
...
+                # for key replace ",", "=", and characters outside the ASCII range 0x20 to 0x7E
+                # for value replace ",", ";", ":" and characters outside the ASCII range 0x20 to 0x7E
+                next_tag = "{}:{};".format(
+                    re.sub("_dd.p.", "t.", re.sub(r",|=|[^\x20-\x7E]+", "_", k)), re.sub(r",|;|:|[^\x20-\x7E]+", "_", v)
+                )
+                if not (len(dd) + len(next_tag)) > 256:
+                    dd += next_tag
+        # remove final ;
```
Comment (mabdinur) [inline ddtrace/context.py:172]: Can we avoid adding the final `;` in the first place

File: ddtrace/context.py [hunk 6]
```diff
@@ -125,9 +132,56 @@ def _traceparent(self):
...
+        if dd:
+            dd = dd[:-1]
+        # If there's a preexisting tracestate we need to update it
+        ts = self._meta.get(_TRACESTATE_KEY)
+        if ts and dd:
+            ts_w_out_dd = re.sub("dd=(.+?),", "", ts)
```
Comment (mabdinur) [inline ddtrace/context.py:178]: Are we guaranteed to have a comma after `dd=(.+?)`? If `ts` exists will dd always come before it?

File: ddtrace/context.py [hunk 7]
```diff
@@ -125,9 +132,56 @@ def _traceparent(self):
...
+            ts = "dd={},{}".format(dd, ts_w_out_dd)
+        elif dd:
+            ts = "dd={}".format(dd)
+        else:
+            ts = ""
```
Comment (mabdinur) [inline ddtrace/context.py:183]: We can void this else statement if we change line 176 to `ts  = self._meta.get(_TRACESTATE_KEY, "")`

File: ddtrace/context.py (hunk 1 above)
Comment (ZStriker19) [reply ddtrace/context.py:135]: We need to pull out the original traceparent so that we can grab the original hexadecimal traceparent trace id. Since we can't convert back from dd trace id -> traceparent trace id, because the dd trace id only represents the lower 64 bits of the 128 digit hexadecimal traceparent id.

File: ddtrace/context.py (hunk 2 above)
Comment (ZStriker19) [reply ddtrace/context.py:156]: Nope, good call. Adding now.

File: ddtrace/context.py (hunk 3 above)
Comment (ZStriker19) [reply ddtrace/context.py]: Yeah I looked for that, but for this one, since it has a specific range of ascii characters, as well as other characters such `,` and `=`, I'm not sure how I'd go about extracting the logic out.

File: ddtrace/context.py (hunk 5 above)
Comment (ZStriker19) [reply ddtrace/context.py:173]: I'm not sure how we could, since we don't know when the final addition of a string is going to be added. I guess one way we could do it is by having each string start with a `;`. Let me try that and see if it works.

File: ddtrace/context.py (hunk 6 above)
Comment (ZStriker19) [reply ddtrace/context.py]: I changed the regex to  `re.sub("dd=(.+?)(?:,|$)", "", ts)` which is what you provided in a different pr and should be more comprehensive.

File: ddtrace/context.py (hunk 5 above)
Comment (ZStriker19) [reply ddtrace/context.py:173]: ah I don't think it would work, as we can't count on the string added before each string to have the `;`.  For example here we can't count on there being a `sampling_priority` value and if there was only an `origin` value
 we'd have `dd=;o:rum` for example, which wouldn't be acceptable.
 ```
        if self.sampling_priority:
            dd += "s:{};".format(self.sampling_priority)
        if self.dd_origin:
            dd += "o:{};".format(self.dd_origin)
```

File: ddtrace/context.py (hunk 1 above)
Comment (mabdinur) [reply ddtrace/context.py:134]: Ahh that makes a lot of sense. I missed your comment bellow

File: ddtrace/context.py [hunk 8]
```diff
@@ -125,9 +131,63 @@ def _traceparent(self):
         # type: () -> str
         if self.trace_id is None or self.span_id is None:
```
Comment (mabdinur) [inline ddtrace/context.py:132]: Since we can now get the trace id from self._meta do we need to update this if statement?

File: ddtrace/context.py [hunk 9]
```diff
@@ -125,9 +131,63 @@ def _traceparent(self):
...
+        usr_id_key = self._meta.get(USER_ID_KEY)
+        if usr_id_key:
+            dd += "t.usr.id:{};".format(re.sub(r",|=|[^\x20-\x7E]+", "_", usr_id_key))
+
+        # grab all other _dd.p values out of meta since we need to propagate all of them
+        for k, v in self._meta.items():
+            if (
+                isinstance(k, six.string_types)
+                and k.startswith("_dd.p")
+                and k not in [SAMPLING_DECISION_TRACE_TAG_KEY, USER_ID_KEY]
```
Comment (mabdinur) [inline ddtrace/context.py:166]: Can we add a comment for why we don't convert these keys? Is there a plan to convert all tags to have the `_dd.p` prefix?

File: ddtrace/context.py (hunk 5 above)
Comment (mabdinur) [reply ddtrace/context.py:176]: Strings are immutable in python so removing a character will generate a new string. It would be a slight performance improvement if we could avoid this but that's not a big deal. We could enumerate over the items in the dictionary and for the last item we can avoid adding ';'.  This logic change would make the for loop harder to read.

tldr: ignore my original suggestion

File: ddtrace/context.py [hunk 10]
```diff
@@ -125,9 +131,63 @@ def _traceparent(self):
...
+                )
+                if not (len(dd) + len(next_tag)) > 256:
+                    dd += next_tag
+        # remove final ;
+        if dd:
+            dd = dd[:-1]
+
+        # If there's a preexisting tracestate we need to update it to preserve other vendor data
+        ts = self._meta.get(_TRACESTATE_KEY, "")
+        if ts and dd:
```
Comment (mabdinur) [inline ddtrace/context.py:182]: Is there a case where the trace key is set and dd="" at this point?  If so should this be?
```suggestion
        if ts:
```

File: ddtrace/context.py [hunk 11]
```diff
@@ -125,9 +131,63 @@ def _traceparent(self):
...
+            ts_w_out_dd = re.sub("dd=(.+?)(?:,|$)", "", ts)
```
Comment (mabdinur) [inline ddtrace/context.py:183]: performance nit: instead of using a regex can check if `ts` == `dd`?

**Review by mabdinur — COMMENTED** [review]
left some small nits, aside from that LGTM

File: ddtrace/context.py (hunk 8 above)
Comment (ZStriker19) [reply ddtrace/context.py:132]: Oh I wasn't aware of that, why do we ever put trace id in meta?

File: ddtrace/context.py (hunk 9 above)
Comment (ZStriker19) [reply ddtrace/context.py:166]: It's because we already converted them above.  Added comment to clarify. I don't think there is a plan to convert all tags to have the `_dd.p` prefix, however the rule is currently that we do propagate all tags with that prefix.  Since we have specifications and rules about the length of sampling decision and user id, we add those one first. Then any other tags that have `_dd.p`, we need to make sure we're not going over the 256 char limit.

File: ddtrace/context.py (hunk 11 above)
Comment (ZStriker19) [reply ddtrace/context.py:183]: I'm not sure I understand, if I don't use regex then I don't know how to cut out the dd section for it to then be replaced. Are you saying that I just check if ts ==dd and if so we'll just return ts and skip unnecessary logic?

Comment (ZStriker19) [reply ddtrace/context.py:183]: Will recheck logic here and add comments

**Comment by majorgreys** [comment]
Summary of changes:

- Update internal interface of `ddtrace.Context` to support traceparent propagation
- Update internal interface of `ddtrace.Context` to support tracestate propagation

File: ddtrace/context.py [hunk 12]
```diff
@@ -123,11 +129,68 @@ def sampling_priority(self, value):
...
+                isinstance(k, six.string_types)
+                and k.startswith("_dd.p")
+                # we've already added sampling decision and user id
+                and k not in [SAMPLING_DECISION_TRACE_TAG_KEY, USER_ID_KEY]
+            ):
+                # for key replace ",", "=", and characters outside the ASCII range 0x20 to 0x7E
+                # for value replace ",", ";", ":" and characters outside the ASCII range 0x20 to 0x7E
+                next_tag = "{}:{};".format(
+                    re.sub("_dd.p.", "t.", re.sub(r",| |=|[^\x20-\x7E]+", "_", k)),
+                    re.sub(r",|;|:|[^\x20-\x7E]+", "_", v),
```
Comment (ZStriker19) [inline ddtrace/context.py:175]: pull out regex and set as constant

File: ddtrace/context.py [hunk 13]
```diff
@@ -123,11 +129,68 @@ def sampling_priority(self, value):
...
+                )
+                if not (len(dd) + len(next_tag)) > 256:
+                    dd += next_tag
+        # remove final ;
+        if dd:
+            dd = dd[:-1]
+
+        # If there's a preexisting tracestate we need to update it to preserve other vendor data
+        ts = self._meta.get(_TRACESTATE_KEY, "")
```
Comment (ZStriker19) [inline ddtrace/context.py:184]: review this logic

File: ddtrace/context.py (hunk 3 above)
Comment (ZStriker19) [reply ddtrace/context.py]: ah nvm, will extract the regex pattern and set as global var

File: ddtrace/context.py [hunk 14]
```diff
@@ -123,11 +130,81 @@ def sampling_priority(self, value):
     @property
     def _traceparent(self):
         # type: () -> str
-        if self.trace_id is None or self.span_id is None:
+        tp = self._meta.get(_TRACEPARENT_KEY)
+        dd_trace_id = self.trace_id
+        span_id = self.span_id
```
Comment (mabdinur) [inline ddtrace/context.py:135]:
```suggestion
        dd_trace_id = self.trace_id
        span_id = self.span_id
```

We can remove these variables. They just add complexity. Let's call `self.span_id` and `self.trace_id` directly

File: ddtrace/context.py [hunk 15]
```diff
@@ -123,11 +130,81 @@ def sampling_priority(self, value):
...
+
+        # determine the trace_id value
+        if tp:
+            # grab the original traceparent trace id, not the converted value
+            trace_id = tp.split("-")[1]
+        elif dd_trace_id:
+            trace_id = "{:032x}".format(dd_trace_id)
+        else:
+            # if we don't have trace id value we can't build a traceparent
+            trace_id = ""
```
Comment (mabdinur) [inline ddtrace/context.py:149]:
```suggestion
            # if we don't have trace id value we can't build a traceparent
            return ""
```

File: ddtrace/context.py [hunk 16]
```diff
@@ -123,11 +130,81 @@ def sampling_priority(self, value):
     @property
     def _traceparent(self):
         # type: () -> str
-        if self.trace_id is None or self.span_id is None:
+        tp = self._meta.get(_TRACEPARENT_KEY)
+        dd_trace_id = self.trace_id
+        span_id = self.span_id
+
+        # if we only have a traceparent then we'll forward it
+        if tp and (span_id is None or dd_trace_id is None):
```
Comment (mabdinur) [inline ddtrace/context.py:138]:
```suggestion
        if tp and (self.span_id is None or self.trace_id is None):
```

File: ddtrace/context.py [hunk 17]
```diff
@@ -123,11 +130,81 @@ def sampling_priority(self, value):
...
+            # grab the original traceparent trace id, not the converted value
+            trace_id = tp.split("-")[1]
+        elif dd_trace_id:
+            trace_id = "{:032x}".format(dd_trace_id)
+        else:
+            # if we don't have trace id value we can't build a traceparent
+            trace_id = ""
+
+        # if we don't have trace id value or a span id value we can't build a traceparent
+        if not trace_id or not span_id:
```
Comment (mabdinur) [inline ddtrace/context.py:152]: if we make the changes above then we don't need to check trace_id again. 

```suggestion
        # if span_id is not set then a valid traceparent can not be set
        if not span_id:
```

File: ddtrace/context.py [hunk 18]
```diff
@@ -27,6 +33,7 @@
         _MetricDictType,  # _metrics
     ]
 
+INVALID_TRACESTATE_TAG_VALUE_CHARS = r",|;|:|[^\x20-\x7E]+"
```
Comment (mabdinur) [inline ddtrace/context.py:36]: small nit: 
We should document this regex so it's easier to maintain. I renamed the constant but the original name is fine too.
```suggestion
# regex matches characters outside the ASCII range 0x20 to 0x7E
TRACESTATE_INVALID_CHARS_REGEX = r",|;|:|[^\x20-\x7E]+"
```

**Review by mabdinur — COMMENTED** [review]
small nits and suggestions. Overall lgtm

File: ddtrace/context.py [hunk 19]
```diff
@@ -123,11 +130,79 @@ def sampling_priority(self, value):
...
+
+        # determine the trace_id value
+        if tp:
+            # grab the original traceparent trace id, not the converted value
+            trace_id = tp.split("-")[1]
+        elif self.trace_id:
+            trace_id = "{:032x}".format(self.trace_id)
+        else:
+            # if we don't have trace id value we can't build a valid traceparent
             return ""
```
Comment (mabdinur) [inline ddtrace/context.py:151]:
```suggestion
        # if we only have a traceparent then we'll forward it
        if self.span_id is None or self.trace_id is None:
            if tp:
                 return tp
           return ""

        # determine the trace_id value
        if tp:
            # grab the original traceparent trace id, not the converted value
            trace_id = tp.split("-")[1]
        else:
            trace_id = "{:032x}".format(self.trace_id)
```

File: ddtrace/context.py [hunk 20]
```diff
@@ -26,7 +32,8 @@
         _MetaDictType,  # _meta
         _MetricDictType,  # _metrics
     ]
-
+# regex matches ",", ";", ":", and characters outside the ASCII range 0x20 to 0x7E
+TRACESTATE_INVALID_CHARS_REGEX = r",|;|:|[^\x20-\x7E]+"
```
Comment (mabdinur) [inline ddtrace/context.py:36]: This should be internal

```suggestion
_TRACESTATE_INVALID_CHARS_REGEX = r",|;|:|[^\x20-\x7E]+"
```

File: ddtrace/context.py [hunk 21]
```diff
@@ -123,11 +130,74 @@ def sampling_priority(self, value):
...
     def _traceparent(self):
         # type: () -> str
-        if self.trace_id is None or self.span_id is None:
+        tp = self._meta.get(_TRACEPARENT_KEY)
+        if self.span_id is None or self.trace_id is None:
+            if tp:
+                # if we only have a traceparent then we'll forward it
+                return tp
+            # if we don't have a span id or trace id value we can't build a valid traceparent
             return ""
```
Comment (majorgreys) [inline ddtrace/context.py:139]:
```suggestion
            # if we only have a traceparent then we'll forward it
            # if we don't have a span id or trace id value we can't build a valid traceparent
            return tp or ""
```

File: ddtrace/context.py [hunk 22]
```diff
@@ -123,11 +130,74 @@ def sampling_priority(self, value):
...
+        if self.dd_origin:
+            dd += "o:{};".format(self.dd_origin)
+        sampling_decision = self._meta.get(SAMPLING_DECISION_TRACE_TAG_KEY)
+        if sampling_decision:
+            # replace characters ",", "=", and characters outside the ASCII range 0x20 to 0x7E
+            dd += "t.dm:{};".format(re.sub(_TRACESTATE_INVALID_CHARS_REGEX, "_", sampling_decision))
+        # since this can change, we need to grab the value off the current span
+        usr_id_key = self._meta.get(USER_ID_KEY)
+        if usr_id_key:
+            dd += "t.usr.id:{};".format(re.sub(_TRACESTATE_INVALID_CHARS_REGEX, "_", usr_id_key))
```
Comment (majorgreys) [inline ddtrace/context.py:166]: Can we pull the tag names out as internal constants we can use elsewhere too?

File: ddtrace/context.py [hunk 23]
```diff
@@ -123,11 +130,74 @@ def sampling_priority(self, value):
...
+                and k not in [SAMPLING_DECISION_TRACE_TAG_KEY, USER_ID_KEY]
+            ):
+                # for key replace ",", "=", and characters outside the ASCII range 0x20 to 0x7E
+                # for value replace ",", ";", ":" and characters outside the ASCII range 0x20 to 0x7E
+                next_tag = "{}:{};".format(
+                    re.sub("_dd.p.", "t.", re.sub(r",| |=|[^\x20-\x7E]+", "_", k)),
+                    re.sub(_TRACESTATE_INVALID_CHARS_REGEX, "_", v),
+                )
+                if not (len(dd) + len(next_tag)) > 256:
+                    dd += next_tag
```
Comment (majorgreys) [inline ddtrace/context.py:183]: This is a good candidate to be pulled out as an internal utility function that can be unit tested.

File: ddtrace/context.py [hunk 24]
```diff
@@ -123,11 +130,74 @@ def sampling_priority(self, value):
...
+        # remove final ;
+        if dd:
+            dd = dd[:-1]
```
Comment (majorgreys) [inline ddtrace/context.py:186]: This could be avoided if we append to a list while going through span tags which we then join once done.

File: ddtrace/context.py [hunk 25]
```diff
@@ -26,7 +32,8 @@
         _MetaDictType,  # _meta
         _MetricDictType,  # _metrics
     ]
-
+# regex matches ",", ";", ":", and characters outside the ASCII range 0x20 to 0x7E
+_TRACESTATE_INVALID_CHARS_REGEX = r",|;|:|[^\x20-\x7E]+"
```
Comment (Kyle-Verhoog) [inline ddtrace/context.py:36]:
```suggestion
_W3C_TRACESTATE_INVALID_CHARS_REGEX = r",|;|:|[^\x20-\x7E]+"
```

without the prefix it sounds like a datadog feature

File: ddtrace/context.py [hunk 26]
```diff
@@ -26,7 +32,8 @@
         _MetaDictType,  # _meta
         _MetricDictType,  # _metrics
     ]
-
+# regex matches ",", ";", ":", and characters outside the ASCII range 0x20 to 0x7E
```
Comment (Kyle-Verhoog) [inline ddtrace/context.py:35]:
```suggestion
```

this comment just echos the actual definition which means that we have to maintain both and likely won't.

File: ddtrace/context.py [hunk 27]
```diff
@@ -123,11 +130,74 @@ def sampling_priority(self, value):
     @property
     def _traceparent(self):
```
Comment (Kyle-Verhoog) [inline ddtrace/context.py:131]: not a huge fan of having these w3c concepts mixed with the `Context` we're implicitly coupling the implementations which will make it hard to refactor one, the other or both in the future.

An alternative might be to just make these standalone functions that take a "`Context`-like" object and operate on that

File: ddtrace/context.py [hunk 28]
```diff
@@ -123,11 +130,74 @@ def sampling_priority(self, value):
...
         # type: () -> str
-        if self.trace_id is None or self.span_id is None:
+        tp = self._meta.get(_TRACEPARENT_KEY)
+        if self.span_id is None or self.trace_id is None:
+            if tp:
+                # if we only have a traceparent then we'll forward it
```
Comment (Kyle-Verhoog) [inline ddtrace/context.py:136]: what does "forward" mean here?

**Review by Kyle-Verhoog — COMMENTED** [review]
Just did a real quick pass.

Can the PR title be a bit more descriptive? It's a bit vague

File: ddtrace/context.py (hunk 23 above)
Comment (mabdinur) [reply ddtrace/context.py:183]: Suggestion:

We can probably split this for loop into two iterations to make easier to read and avoid removing the final ';':

Iteration 1: Use a for loop to get all the tags that start with `_dd.p`, replace invalid characters and store the tags in an array (ex: ["t.dm:-4", "t.usr.id:baz64", ....]). 

Iteration 2: Use [string.join(';') ](https://www.programiz.com/python-programming/methods/string/join) to convert the array into a string. This way you don't need to remove the final ';' from the string.

File: ddtrace/context.py (hunk 28 above)
Comment (ZStriker19) [reply ddtrace/context.py]: It means propagate, changed wording.

File: ddtrace/context.py (hunk 22 above)
Comment (ZStriker19) [reply ddtrace/context.py]: We cam with origin and sampling priority. User id key and sampling decision maker don't get used elsewhere.

File: ddtrace/internal/utils/formats.py [hunk 29]
```diff
@@ -160,3 +166,26 @@ def stringify_cache_args(args):
             break
 
     return " ".join(out)
+
+
+def _w3c_format_unknown_propagated_tags(current_tags, potential_tags):
```
Comment (mabdinur) [inline ddtrace/internal/utils/formats.py:171]: nit: I am not sure if this module is the best place for this function. Maybe we can should move it to `ddtrace/internal/utils/http.py`?
@Kyle-Verhoog  thoughts?

File: ddtrace/context.py [hunk 30]
```diff
@@ -123,11 +131,57 @@ def sampling_priority(self, value):
...
+        if self.dd_origin:
+            # the origin value has specific values that are allowed.
+            dd.append("{}:{}".format(_W3C_TRACESTATE_ORIGIN_KEY, re.sub(r",|;|=|[^\x20-\x7E]+", "_", self.dd_origin)))
+        sampling_decision = self._meta.get(SAMPLING_DECISION_TRACE_TAG_KEY)
+        if sampling_decision:
+            dd.append("t.dm:{}".format(re.sub(_W3C_TRACESTATE_INVALID_CHARS_REGEX, "_", sampling_decision)))
+        # since this can change, we need to grab the value off the current span
+        usr_id_key = self._meta.get(USER_ID_KEY)
+        if usr_id_key:
+            dd.append("t.usr.id:{}".format(re.sub(_W3C_TRACESTATE_INVALID_CHARS_REGEX, "_", usr_id_key)))
```
Comment (mabdinur) [inline ddtrace/context.py:165]: nit: we can probably move this to a utils function too and name it `_w3c_format_known_propagated_tags`?

File: tests/tracer/test_utils.py [hunk 31]
```diff
@@ -398,3 +400,71 @@ def _(a, b=None, *args, **kwargs):
         @callonce
         def _():
             yield 42
+
+
+@pytest.mark.parametrize(
```
[the remark this answers is not in the document]
Comment (ZStriker19) [reply tests/tracer/test_utils.py:405]: ah it's because they depended on dict order, fixing that.

File: ddtrace/internal/utils/formats.py (hunk 29 above)
Comment (ZStriker19) [reply ddtrace/internal/utils/formats.py:171]: I wasn't sure either, I thought about http.py as well but didn't want it to necessarily be only for http propagation? Since we use it in other instances, such as for your dbm work.

File: ddtrace/internal/utils/formats.py [hunk 32]
```diff
@@ -160,3 +166,26 @@ def stringify_cache_args(args):
...
+        ):
+            # for key replace ",", "=", and characters outside the ASCII range 0x20 to 0x7E
+            # for value replace ",", ";", ":" and characters outside the ASCII range 0x20 to 0x7E
+            next_tag = "{}:{}".format(
+                re.sub("_dd.p.", "t.", re.sub(r",| |=|[^\x20-\x7E]+", "_", k)),
+                re.sub(_W3C_TRACESTATE_INVALID_CHARS_REGEX, "_", v),
+            )
+            # we need to keep the total length under 256 char
+            current_tags_len += len(next_tag)
+            if not current_tags_len > 256:
```
Comment (mabdinur) [inline ddtrace/internal/utils/formats.py:189]: Question: Should we add a debug/warning log here? Is this behavior documented?

Comment (ZStriker19) [reply ddtrace/internal/utils/formats.py:189]: yeah that's a good idea

File: ddtrace/context.py (hunk 27 above)
Comment (ZStriker19) [reply ddtrace/context.py:129]: @Kyle-Verhoog  would this be a blocker?

File: ddtrace/internal/utils/formats.py (hunk 29 above)
Comment (ZStriker19) [reply ddtrace/internal/utils/formats.py:173]: I'll put in http.py

File: ddtrace/context.py (hunk 27 above)
Comment (majorgreys) [reply ddtrace/context.py:129]: I agree we should reconsider the coupling of w3c headers and the context object but would suggest that be done as a follow-up refactoring to this change.

**Review by majorgreys — DISMISSED** [review]

File: ddtrace/context.py (hunk 27 above)
Comment (Kyle-Verhoog) [reply ddtrace/context.py:129]: :+1: i'm ok to not block on it

File: ddtrace/context.py [hunk 33]
```diff
@@ -10,7 +11,11 @@
 from .constants import USER_ID_KEY
 from .internal.compat import NumericType
 from .internal.compat import PY2
+from .internal.constants import _TRACEPARENT_KEY
+from .internal.constants import _TRACESTATE_KEY
 from .internal.logger import get_logger
+from .internal.utils.http import w3c_format_known_propagated_tags
+from .internal.utils.http import w3c_format_unknown_propagated_tags
```
Comment (mabdinur) [inline ddtrace/context.py:18]: we should avoid leaking this function in the public api. Sorry I should've noticed this earlier:

```suggestion
from .internal.utils.http import w3c_format_known_propagated_tags as _w3c_format_known_propagated_tags
from .internal.utils.http import w3c_format_unknown_propagated_tags as _w3c_format_unknown_propagated_tags
```

**Review by mabdinur — COMMENTED** [review]
One last nit

**Review by mabdinur — APPROVED** [review]

File: ddtrace/internal/constants.py [hunk 34]
```diff
@@ -8,5 +8,8 @@
     PROPAGATION_STYLE_B3,
     PROPAGATION_STYLE_B3_SINGLE_HEADER,
 )
-
+_TRACESTATE_KEY = "tracestate"
+_TRACEPARENT_KEY = "traceparent"
+_W3C_TRACESTATE_ORIGIN_KEY = "o"
+_W3C_TRACESTATE_SAMPLING_PRIORITY_KEY = "s"
```
Comment (majorgreys) [inline ddtrace/internal/constants.py:14]: Any reason these are being added with underscore prefix in an internal module? Also, we could add the `W3C_` prefix for clarity.

```suggestion
W3C_TRACESTATE_KEY = "tracestate"
W3C_TRACEPARENT_KEY = "traceparent"
W3C_TRACESTATE_ORIGIN_KEY = "o"
W3C_TRACESTATE_SAMPLING_PRIORITY_KEY = "s"
```

File: ddtrace/internal/utils/formats.py [hunk 35]
```diff
@@ -19,7 +19,6 @@
 VALUE_TOO_LONG_MARK = "..."
 CMD_MAX_LEN = 1000
 
-
```
Comment (majorgreys) [inline ddtrace/internal/utils/formats.py:22]: ?

File: ddtrace/context.py [hunk 36]
```diff
@@ -123,11 +128,45 @@ def sampling_priority(self, value):
...
         sampled = 1 if self.sampling_priority and self.sampling_priority > 0 else 0
-        return "00-{:032x}-{:016x}-{:02x}".format(self.trace_id, self.span_id, sampled)
+        return "00-{}-{:016x}-{:02x}".format(trace_id, self.span_id, sampled)
+
+    @property
+    def _tracestate(self):
+        # type: () -> str
+        dd = _w3c_format_known_propagated_tags(self)
+
+        # get and format all other _dd.p values out of meta since we need to propagate all of them
```
Comment (majorgreys) [inline ddtrace/context.py:152]: Why are these two functions split here though we seem to be calling these two functions together?

Comment (ZStriker19) [reply ddtrace/context.py:152]: Because they were called out at different times as pieces of logic that could be broken out. I'll combine them and change the tests.

File: ddtrace/internal/utils/formats.py [hunk 37]
```diff
@@ -19,7 +19,6 @@
...
 T = TypeVar("T")
```
Comment (ZStriker19) [inline ddtrace/internal/utils/formats.py:22]:
```suggestion

T = TypeVar("T")
```

**Review by mabdinur — DISMISSED** [review]
🚢

**Review by majorgreys — APPROVED** [review]

**Review by mabdinur — DISMISSED** [review]

**Review by majorgreys — APPROVED** [review]

**Review by mabdinur — APPROVED** [review]
